server.js

```diff
@@ -1,33 +1,127 @@
-import express from 'express'
-import bodyParser from 'body-parser'
-import cors from 'cors'
-
-// If you're using one of our datasets, uncomment the appropriate import below
-// to get started!
-// 
-// import goldenGlobesData from './data/golden-globes.json'
-// import avocadoSalesData from './data/avocado-sales.json'
-// import booksData from './data/books.json'
-// import netflixData from './data/netflix-titles.json'
-// import topMusicData from './data/top-music.json'
+import express from 'express';
+import bodyParser from 'body-parser';
+import cors from 'cors';
+import booksData from './data/books.json';
 
 // Defines the port the app will run on. Defaults to 8080, but can be 
 // overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
-const port = process.env.PORT || 8080
-const app = express()
+// PORT=9000 npm start
+const port = process.env.PORT || 8080;
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+app.use(cors());
+app.use(bodyParser.json());
+
+const listEndpoints = require('express-list-endpoints');
 
 // Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  if(!res) {
+    setTimeout(() => {
+      res
+      .status(404)
+      .send({ Error: 'Sorry, a problem occured try again later' });
+    }, 8000);
+  }
+  else res.send(listEndpoints(app));
+});
+
+// Route of books array, pagination as default
+app.get('/books', (req, res) => {
+  const { sort, author, title } = req.query;
+  let booksList = booksData;
+  const totalOfBooks = booksList.length ;
+
+  //sort by rating
+  if(sort === "rating_dsc") {
+    booksList = booksList.sort((a, b) => (b.average_rating - a.average_rating));
+  } else if (sort === "rating_asc") {
+    booksList = booksList.sort((a, b) => (a.average_rating - b.average_rating));
+  };
+  
+  //filter by author
+  if (author) {
+    booksList = booksList.filter((item) => item.authors.toString().toLowerCase().includes(author.toLowerCase()));
+  };
+  
+  //filter by title
+  if (title) {
+    booksList = booksList.filter((item) => item.title.toString().toLowerCase().includes(title.toLowerCase()));
+  };
+
+  //PAGINATION limit of 20 results per page
+  const page = req.query.page ? req.query.page -1 : 0;
+  const pageSize = req.query.pageSize ?? 20;
+  const startIndex = page * pageSize;
+  const endIndex = startIndex + (+pageSize);
+  const booksListPage = booksList.slice(startIndex, endIndex);
+  const returnObject = { 
+    totalNumberOfBooks: totalOfBooks,
+    totalNumberOfPages: Math.ceil(booksList.length / pageSize), 
+    booksReturned: pageSize, 
```

Comment:
The name `booksReturned` is a little confusing - if I query:
`/books?title=harry`
and get 9 responses but it says `"booksReturned": 20` I would expect page 2 to hold 11 books 🤔